Add support for batch_job_definition and batch_job_queue

ashinohara opened the pull request:
## Description
Adding resources `aws_batch_job_definition` and `aws_batch_job_queue`.

Additionally, I found an issue with some of the tests for `aws_batch_compute_environment` which seems to be a condition where the `aws_iam_role_policy_attachment` gets deleted before the compute environment and causes the `aws_batch_compute_environment ` to be a dangling resource.  I have added a note in the documentation about it, and have also fixed the tests by adding a `depends_on` link in the `aws_batch_compute_environment ` to the `aws_iam_role_policy_attachment`.

## Tests
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSBatchJobDefinition'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSBatchJobDefinition -timeout 120m
=== RUN   TestAccAWSBatchJobDefinition
--- PASS: TestAccAWSBatchJobDefinition (14.26s)
=== RUN   TestAccAWSBatchJobDefinitionUpdateForcesNewResource
--- PASS: TestAccAWSBatchJobDefinitionUpdateForcesNewResource (22.15s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	36.442s
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSBatchJobQueue'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSBatchJobQueue -timeout 120m
=== RUN   TestAccAWSBatchJobQueue
--- PASS: TestAccAWSBatchJobQueue (89.57s)
=== RUN   TestAccAWSBatchJobQueueUpdate
--- PASS: TestAccAWSBatchJobQueueUpdate (104.32s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	193.919s
```

radeksimko commented on aws/resource_aws_batch_job_definition.go:
```diff
@@ -0,0 +1,209 @@
+		log.Printf("[DEBUG] Job Definition %q is already gone", arn)
+		return nil, nil
+	case numJobDefinitions == 1:
+		return resp.JobDefinitions[0], nil
```
Shouldn't we also check the status of the definition here, since we treat `INACTIVE` definition as "deleted" I think it should be reflected here too, unless there's a strong reason not to do it.

radeksimko commented on aws/resource_aws_batch_job_definition_test.go:
```diff
@@ -0,0 +1,264 @@
+func testAccCheckJobDefinitionRecreated(t *testing.T,
+	before, after *batch.JobDefinition) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if *before.Revision == *after.Revision {
```
That's quite clever! 😄  👍

radeksimko commented on aws/resource_aws_batch_job_definition_test.go:
```diff
@@ -0,0 +1,264 @@
+		conn := testAccProvider.Meta().(*AWSClient).batchconn
+		js, err := getJobDefinition(conn, rs.Primary.Attributes["arn"])
+		if err == nil {
+			if *js.Status != "INACTIVE" {
```
Really a nitpick, but positive conditionals tend to be easier to read, i.e. `*js.Status == "ACTIVE"`.

radeksimko commented on aws/resource_aws_batch_job_queue.go:
```diff
@@ -0,0 +1,206 @@
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Is there any reason why `arn` should be Optional? I don't see it being used anywhere in the CRUD other than in `Set()` which makes me think it should be Computed-only.

radeksimko commented on aws/resource_aws_batch_job_queue.go:
```diff
@@ -0,0 +1,206 @@
+		Priority:                aws.Int64(int64(d.Get("priority").(int))),
+		State:                   aws.String(d.Get("state").(string)),
+	}
+	_, err := conn.UpdateJobQueue(updateInput)
```
Shouldn't we also wait here for the job until it goes from `batch.JQStatusUpdating` to `batch.JQStatusValid`?

radeksimko commented on aws/resource_aws_batch_job_queue.go:
```diff
@@ -0,0 +1,206 @@
+	return nil
+}
+
+func createComputeEnvironmentOrder(d *schema.ResourceData) (envs []*batch.ComputeEnvironmentOrder) {
```
Maybe a nitpick, but the interface here seems to be quite greedy - any reason we can't just pass an `[]interface{}` and call `d.Get()` outside of this function since it only needs access to a single field, not all ResourceData?

radeksimko commented on aws/resource_aws_batch_job_queue.go:
```diff
@@ -0,0 +1,206 @@
+			return nil
+		}
+		return resource.RetryableError(err)
+	})
```
Is there any particular reason why this should fail and/or why we should retry? Can we retry on a specific error code, instead of _any_ error?

radeksimko replied:
Also if we really need the retry logic, then we should also catch the `err` coming from it - right now it seems to be shadowed on line 152.

ashinohara replied:
I was using it to wait until the Job Queue was disabled, but I think a more elegant solution is to use `WaitForState`.  I am working on that change.

radeksimko commented on aws/validators.go:
```diff
@@ -1420,10 +1420,10 @@ func validateSsmParameterType(v interface{}, k string) (ws []string, errors []er
 	value := v.(string)
 	if !(regexp.MustCompile(`^[A-Za-z0-9_]*$`).MatchString(value) && len(value) <= 128) {
-		errors = append(errors, fmt.Errorf("computeEnvironmentName must be up to 128 letters (uppercase and lowercase), numbers, and underscores."))
+		errors = append(errors, fmt.Errorf("Name must be up to 128 letters (uppercase and lowercase), numbers, and underscores."))
```
I have raised https://github.com/terraform-providers/terraform-provider-aws/pull/1760/files to address the generic-ness of this function - I agree it can be still renamed, but either way - do you mind rebasing?

radeksimko commented on website/docs/r/batch_job_definition.html.markdown:
````diff
@@ -0,0 +1,78 @@
+
+```hcl
+resource "aws_batch_job_definition" "test" {
+	name = "tf-test-batch-job-definition"
````
I think we should replace hyphens with underscores here, per validation function.

ashinohara commented on aws/resource_aws_batch_job_definition.go:
```diff
@@ -0,0 +1,209 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				StateFunc: func(v interface{}) string {
```
Yes, that was the intent.  Thanks for pointing me to that function.

ashinohara commented:
@radeksimko Thank you very much for your insightful review.  I believe I have addressed all of your comments and made sure that tests are still passing.  Let me know if there is anything else you would like me to change.

radeksimko commented on aws/resource_aws_batch_job_definition.go:
```diff
@@ -0,0 +1,212 @@
+			},
+			"revision": {
+				Type:     schema.TypeInt,
+				Optional: true,
```
The `revision` here is still `Optional`, but I will take care of that prior to merging.

pat-nomalab commented:
Hi,
thanks for this feature
I've managed to create batch compute env with job queue link to it, but I have an issue when i'm trying to upgrade compute env definition "Cannot delete, found existing JobQueue relationship"
